### PR TITLE
tsserver: get candidates from <K extends keyof Foo> (fix #12536)

### LINKED ISSUE #12536: string literal completion for `<K extends keyof Foo>` does not work
**TypeScript Version:**  nightly (2.2.0-dev.20161128)

**Code**

```ts
interface Foo { foo: string; }
function ok(a: keyof Foo) {};
function ng<K extends keyof Foo>(a: K) {};

ok("/*position 1*/");
ng("/*position 2*/");
```

**Expected behavior:**
tsserver returns `foo` as completion candidate for both positions.

**Actual behavior:**
tsserver returns no candidate for position 2.

### PR DESCRIPTION
**Pull request by wonderful-panda**
Fixes #12536

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -244,6 +244,9 @@ namespace ts.Completions {
         }
 
         function addStringLiteralCompletionsFromType(type: Type, result: CompletionEntry[]): void {
+            if (type && type.flags & TypeFlags.TypeParameter) {
```
Comment (DanielRosenwasser) [inline src/services/completions.ts]: If you get the type constraint and the type is a `keyof` type, where is that handled below?

File: src/services/completions.ts [hunk 2]
```diff
@@ -244,6 +244,9 @@ namespace ts.Completions {
         }
 
         function addStringLiteralCompletionsFromType(type: Type, result: CompletionEntry[]): void {
+            if (type && type.flags & TypeFlags.TypeParameter) {
+                type = (<TypeParameter>type).constraint;
```
Comment (mhegazy) [inline src/services/completions.ts]: you should use `checker::getApparentType` here instead. it does some more than just checking the constraint.

Comment (mhegazy) [inline src/services/completions.ts]: to do that, you will need to expose this method on TypeChecker interface.

**Comment by wonderful-panda** [comment]
Thank you for the review. I've pushed new version.

**Pull request merged**